engine-precompiles/src/alt_bn256.rs

```diff
@@ -42,21 +42,144 @@ mod consts {
 
     /// Pair element length.
     pub(super) const PAIR_ELEMENT_LEN: usize = 192;
+
+    pub(super) const SCALAR_PART_LEN: usize = SCALAR_LEN / 2;
+
+    /// Size of BN scalars.
+    pub(super) const SCALAR_LEN: usize = 32;
+
+    /// Half the size of a point size.
+    pub(super) const POINT_PART_LEN: usize = POINT_LEN / 2;
+
+    /// Size of BN points.
+    pub(super) const POINT_LEN: usize = 64;
+
+    /// Size of BN pairs.
+    pub(super) const POINT_PAIR_LEN: usize = 128;
+
+    /// Output length.
+    pub(super) const OUTPUT_LEN: usize = 64;
+}
+
+#[cfg(feature = "contract")]
+trait SdkEncode {
+    type Encoded;
+
+    fn sdk_encode(self) -> Self::Encoded;
+}
+
+#[cfg(feature = "contract")]
+impl SdkEncode for bn::Fr {
+    type Encoded = [u8; consts::SCALAR_LEN];
+
+    fn sdk_encode(self) -> Self::Encoded {
+        let (low, high) = {
+            let [low_int, high_int] = self.into_u256().0;
+            (low_int.to_le_bytes(), high_int.to_le_bytes())
+        };
+
+        let mut bytes = [0u8; consts::SCALAR_LEN];
+        bytes[0..consts::SCALAR_PART_LEN].copy_from_slice(&low);
+        bytes[consts::SCALAR_PART_LEN..consts::SCALAR_PART_LEN * 2].copy_from_slice(&high);
+        bytes
+    }
+}
+
+#[cfg(feature = "contract")]
+impl SdkEncode for bn::G1 {
+    type Encoded = [u8; consts::POINT_LEN];
+
+    fn sdk_encode(self) -> Self::Encoded {
+        bn::AffineG1::from_jacobian(self)
+            .map(|p| {
+                // It is possible to use sdk_encode for Fr here but, this is
+                // just a small optimisation to avoid additional allocations.
```

Comment:
Sounds like a premature optimization to me. I would rather see the `sdk_encode` call here because it avoids code duplication and makes the code here a little easier to read. It's also not clear to me how much of an optimization this is in the first place. Rust arrays are not heap-allocated, they live on the stack, so are much less overhead to create than a `Vec`.

Reply:
I was just quite a bit worried about how much were being used, and wasn't sure in a wasm context if this actually would cause an issue.

Reply:
Actually, even then the comment didn't make a whole lot of sense given the actual use. I probably intended to do something else, but got a different result.